From 7426755612368dde9f104d7de4aaae060a6bc417 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:36:29 -0400
Subject: [ethernet] i40e: Fix a bug in the update logic for FDIR SB filter

Message-id: <1409658042-22689-29-git-send-email-sassmann@redhat.com>
Patchwork-id: 90223
O-Subject: [RHEL7.1 PATCH 028/281] i40e: Fix a bug in the update logic for FDIR SB filter.
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Anjali Singhai Jain <anjali.singhai@intel.com>

The update filter logic was causing a kernel panic in the original code.
We need to compare the input set to decide whether or not to delete a
filter since we do not have a hash stored. This new design helps fix the issue.

Change-ID: I2462b108e58ca4833312804cda730b4660cc18c9
Signed-off-by: Anjali Singhai Jain <anjali.singhai@intel.com>
Signed-off-by: Catherine Sullivan <catherine.sullivan@intel.com>
Tested-by: Kavindya Deegala <kavindya.s.deegala@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 43fddb7576fbd543502b01f0c09a3a4171f7e038)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_ethtool.c b/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
index cb6fc95..6f4188a 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
@@ -1357,6 +1357,24 @@ static int i40e_set_rss_hash_opt(struct i40e_pf *pf, struct ethtool_rxnfc *nfc)
 }
 
 /**
+ * i40e_match_fdir_input_set - Match a new filter against an existing one
+ * @rule: The filter already added
+ * @input: The new filter to comapre against
+ *
+ * Returns true if the two input set match
+ **/
+static bool i40e_match_fdir_input_set(struct i40e_fdir_filter *rule,
+          struct i40e_fdir_filter *input)
+{
+ if ((rule->dst_ip[0] != input->dst_ip[0]) ||
+     (rule->src_ip[0] != input->src_ip[0]) ||
+     (rule->dst_port != input->dst_port) ||
+     (rule->src_port != input->src_port))
+  return false;
+ return true;
+}
+
+/**
  * i40e_update_ethtool_fdir_entry - Updates the fdir filter entry
  * @vsi: Pointer to the targeted VSI
  * @input: The filter to update or NULL to indicate deletion
@@ -1391,11 +1409,10 @@ static int i40e_update_ethtool_fdir_entry(struct i40e_vsi *vsi,
 
  /* if there is an old rule occupying our place remove it */
  if (rule && (rule->fd_id == sw_idx)) {
-  if (!input || (rule->fd_id != input->fd_id)) {
-   cmd->fs.flow_type = rule->flow_type;
-   err = i40e_add_del_fdir_ethtool(vsi, cmd, false);
-  }
-
+  if (input && !i40e_match_fdir_input_set(rule, input))
+   err = i40e_add_del_fdir(vsi, rule, false);
+  else if (!input)
+   err = i40e_add_del_fdir(vsi, rule, false);
   hlist_del(&rule->fdir_node);
   kfree(rule);
   pf->fdir_pf_active_filters--;
-- 
1.7.1